patchSourcetimePath(Bash|Csh|Fish|Posix): init; root: fix thisroot.* setxrd.*

ShamrockLee opened the pull request:
###### Description of changes
*   Add setup hook `patchSourcetimePath(Bash|Csh|Fish|Posix)` to add extra bin paths to `PATH` during the time the script is sourced, and clean them up afterward. Several test cases are also added to ensure that those hooks do their jobs.

*   Use the above setup hooks to patch thisrooot.* and setxrd.* for ROOT, so that the builder won't complain about command not found when sourcing thisroot.sh and building packages that depends on ROOT. Most of the files added are part of the test cases of the hooks.

*   Require #186323 (will rebase this PR when that one gets merged)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

veprbl commented:
May have to patch llvm with https://github.com/root-project/root/commit/a9c961cf4613ff1f0ea50f188e4a4b0eb749b17d

veprbl commented on pkgs/applications/science/misc/root/default.nix:
```diff
@@ -38,15 +38,16 @@
 , CoreSymbolication
 , OpenGL
 , noSplash ? false
+, swVersPatch ? ./sw_vers.patch
```
This is a bit unusual

ShamrockLee replied:
The motivation is to make version (source) overriding easier.

I have a flake.nix in my analysis project that tracks the master branch of root-project/root. The ./sw_vers.patch is a moving target that needs to be regenerated from time to time.

I acknowledge that it's kinda "unusual". I would change it back if need be.

veprbl replied:
I also often maintain custom overrides. In my experience with packaging bleeding edge version, I often just end up dropping all values from the original `patches` attribute.

The problem with doing one-off interfaces is that there will be other packages you might want to override and your existing experience around `swVersPatch` will not translate to that.

ShamrockLee replied:
I see.

ShamrockLee commented:
> May have to patch llvm with [root-project/root@a9c961c](https://github.com/root-project/root/commit/a9c961cf4613ff1f0ea50f188e4a4b0eb749b17d)

@veprbl Should the patch goes directly to the `llvm_9` package in Nixpkgs, or should we customize that LLVM to inject into the 'buildInputs' here?

ShamrockLee committed:
root: wrap the executable and patch thisroot.* setxrd.*

veprbl commented on pkgs/applications/science/misc/root/default.nix:
```diff
@@ -178,6 +195,54 @@ stdenv.mkDerivation rec {
+          which-p "which -p"
+      done
+    }
+    patchCommandInplace "$out/bin/thisroot.sh" \
+      "grep" "${gnugrep}/bin/grep" \
+      "manpath" "${man}/bin/manpath" \
+      "man" "${man}/bin/man" \
+      "ps" "${procps}/bin/ps" \
+      "sed" "${gnused}/bin/sed"
```
Is there a reason why
```bash
substituteInPlace "$out/bin/thisroot.sh" \
  --replace "grep" "${gnugrep}/bin/grep" \
  # ...
```
won't work?

ShamrockLee replied:
As a general approach, direct substitution may cause issue of replacing longer words that contains the sub-string. For example, `manpath` might be affected by the substitution of `man`.

veprbl replied:
That is why you substitute "manpath" first?

ShamrockLee replied:
@veprbl I didn't thought about the order (which is roughly alphabetic) actually. In the `patchCommandInplace` bash function, I use four sed commands to simulate the regex lookaround and ensure that the whole word is being replaced.

veprbl replied:
Alternative idea: put all this stuff to `propagatedBuildInputs`

veprbl commented:
```
error: builder for '/nix/store/rf3vrqywrdiy029hymva2y1s9jpm3dn5-llvm-9.0.1.drv' failed with exit code 1;
       last 10 log lines:
       > --------------------------
       > |diff --git a/llvm/tools/llvm-readobj/ELFDumper.cpp b/llvm/tools/llvm-readobj/ELFDumper.cpp
       > |index 4e1cb7d544e7..972d41ea104b 100644
       > |--- a/llvm/tools/llvm-readobj/ELFDumper.cpp
       > |+++ b/llvm/tools/llvm-readobj/ELFDumper.cpp
       > --------------------------
       > File to patch:
       > Skip this patch? [y]
       > Skipping patch.
       > 1 out of 1 hunk ignored
       For full logs, run 'nix log /nix/store/rf3vrqywrdiy029hymva2y1s9jpm3dn5-llvm-9.0.1.drv'.
```

veprbl commented:
Please try to use `fetchpatch`, it has some parameters to modify paths and select which files to patch:
https://github.com/NixOS/nixpkgs/blob/1f117e37a20d491ae1cbde9cac114f6dbd7f217e/pkgs/build-support/fetchpatch/default.nix#L13-L16

ShamrockLee committed:
patchSourcetimePath(Bash|Csh|Fish|Posix): init

with test cases

ShamrockLee committed:
trivial-builders.nix: Add input argument `passthru` to makeSetupHook

One significant use case is adding `passthru.tests` to setup-hooks,
and help increase test coverage for mission-critical setup-hooks.

As `meta`, `passthru` doesn't go into the build script directly.
However, passing an empty set to `passthru` breaks nixpkgs-review
and OfBorg tests, so pass it only when specified.

ShamrockLee committed:
root: format expression with nixpkgs-format
